### DISCUSSION TITLE
implementation of seurat DotPlot function

### DISCUSSION THREAD

**Discussion by NoemieL** [GreenleafLab/ArchR#882] [Feature Requests]
Hi,
Could it be possible to have a similar function as the one proposed in seurat with DotPlot (https://satijalab.org/seurat/archive/v3.0/visualization_vignette.html) that provide a graph with the percentage of cells expressing the feature and the average expression level in each cluster. Could be nice to have this available for GeneScore and GeneIntegration.
![image](https://user-images.githubusercontent.com/26926109/125362728-9f8fce00-e32c-11eb-9ac6-7a2630ef0007.png)
Thanks

**Comment by sylestiel** [comment]
Were you able to figure this out?

**Reply by NoemieL** [reply]
Yes, you could find my script here https://github.com/NoemieL/bubble-plot-ArchR
Please let me know if something is not clear.

**Reply by ZayaAyush** [reply]
Thank you, NoemieL. Can you please tell me how to change the color in a bubble plot?

**Comment by sylestiel** [comment]
Thanks NoemieL for the script. 

I have a few questions:
How did you generate the gene_list?

Also can the same be used to plot using the gene score matrix? If so how should I modify the script.

**Reply by NoemieL** [reply]
For gene_list, you could do gene_list=c("TP53","BCL2","TGFB") or take the list of gene from ArchR

To use the gene score : genescore= getMatrixFromProject(ArchRProj, useMatrix="GeneScoreMatrix")

**Reply by sylestiel** [reply]
Thanks for your prompt reply.

**Reply by matteozoia4** [reply]
Dear Noemie many thanks for your script!

I encounter an error during bubble dataframe creation based on **GeneExpressionMatrix**. 

Dataframe creation :
```
bubble_plot_info=data.frame()
for(i in gene_list){
  for(k in 1:length(unique(geneinteinfo2$Clusters))){
    a=nrow(bubble_plot_info)
    l=unique(geneinteinfo2$Clusters)[k]
    bubble_plot_info[a+1,"gene_name"]=i
    bubble_plot_info[a+1,"Clusters"]=l
    eval(parse(text=(paste("bubble_plot_info[",a,"+1,'pct_exp']=(length(geneinteinfo2[(geneinteinfo2$",i,">0 & geneinteinfo2$Clusters=='",l,"'),'",i,"'])/nrow(geneinteinfo2[geneinteinfo2$Clusters=='",l,"',]))*100", sep=""))))
    eval(parse(text=(paste("bubble_plot_info[",a,"+1,'avg_exp']=mean(geneinteinfo2[geneinteinfo2$",i,">0 & geneinteinfo2$Clusters=='",l,"','",i,"'])", sep=""))))
  }
}
```
Error :
```
In mean.default(geneinteinfo2[geneinteinfo2$HCN4 > 0 &  ... :
  argument is not numeric or logical: returning NA
```
How could I overcome this error?

Many thanks for he help!

MZ

**Reply by NoemieL** [reply]
Hi Matteo,
Does it work with other genes? Are you sure your HCN4 gene is expressed in your cells? Or is it present in your GeneExpressionMatrix?
If not, your values may not be numerical. It's hard to tell since you didn't show how your data is. Could you give more information about what you have done before and how your data looks like?
Thanks

**Reply by NoemieL** [reply]
I looked in my data and your gene is not present in the GeneExpressionMatrix, I also tried the aliases BRGDA8, EIG18, and SSS2. I think this is the problem and the only way to get it is to find a way to not exclude its region on your scRNAseq data.

**Comment by sylestiel** [comment]
Any suggestions on how to specify the order in which the genes are plotted?

Also how to are used the predicted group names of clusters post integration?

**Reply by NoemieL** [reply]
You can adapt the ggplot in your convenience as follow https://www.r-graph-gallery.com/267-reorder-a-variable-in-ggplot2.html

If I understoud well your question, you could change the code here to adapt based on what info you want, especially the **ArchRProj$**Clusters****
for(i in rownames(geneinteinfo2)){
  geneinteinfo2[i,"Clusters"]=**ArchRProj$**Clusters****[which(ArchRProj$cellNames==i)]
  geneinteinfo2[i,"Samples"]=**ArchRProj$**Sample****[which(ArchRProj$cellNames==i)]
}

**Reply by sylestiel** [reply]
Thanks!

**Reply by sylestiel** [reply]
@NoemieL 

How can I modify the script to incorporate scaled values instead of avg_expression?

**Reply by NoemieL** [reply]
I am not sure I understand what you want to do. If you want to scaled the avg_expression you could create a new column in geneinteinfo2 scale(geneinteinfo2$avg_expression) and then select this column as x in the ggplot.

**Reply by sylestiel** [reply]
You are understanding me correctly. However, I'm not familiar with scripting. So don't know how to incorporate scaling on average expression for the dot or bubble plot. 
I'm getting the following error:

colnames(geneinteinfo2)=scale(geneinteinfo2$avg_expression)
Error in array(x, c(length(x), 1L), if (!is.null(names(x))) list(names(x),  : 
  'data' must be of a vector type, was 'NULL'

**Reply by NoemieL** [reply]
Sorry, it is on bubble_plot_info and not on geneinteinfo2 that you need to add a column.
After creating your bubble_plot_info, do:
bubble_plot_info[is.na(bubble_plot_info$avg_exp),"avg_exp"]=0 #to replace NA value by 0

for(i in gene_list){
bubble_plot_info[which(bubble_plot_info$gene_name==i),'scale(avg_exp)']=scale(bubble_plot_info[which(bubble_plot_info$gene_name==i),'avg_exp'])
} # to scale our data for each gene

and change in the ggplot color = "scale(avg_exp)"

**Reply by sylestiel** [reply]
Appreciate it!! It worked.

**Comment by sylestiel** [comment]
@NoemieL 

I would like to change the script for percent expression to reflect percent of cells in all clusters and not just in individual clusters. How may I incorporate that into your script for dot plots?

**Reply by NoemieL** [reply]
Sorry I was busy. If you want the percentage of cells that expressed the gene without taking into account the clustering, you just have to remove this geneinteinfo2$Clusters=='",l,"' that is the condition of being in a specific cluster.

I would do another data.frame as follow:

bubble_plot_info=data.frame()
for(i in gene_list){
    a=nrow(bubble_plot_info)
    bubble_plot_info[a+1,"gene_name"]=i
    eval(parse(text=(paste("bubble_plot_info[",a,"+1,'pct_exp_total']=(length(geneinteinfo2[(geneinteinfo2$",i,">0 ,'",i,"'])/nrow(geneinteinfo2))*100", sep=""))))
    eval(parse(text=(paste("bubble_plot_info[",a,"+1,'avg_exp']=mean(geneinteinfo2[geneinteinfo2$",i,">0 ,'",i,"'])", sep=""))))
}

**Reply by sylestiel** [reply]
Thank you!!!

**Comment by matteozoia4** [comment]
Dear Noemie,


many thanks for your tempestive answer!

I am looking through this, I may indeed have set accidentally to uppercases the gene names, since working with mouse this could be definitely the issue for matching.

As well I need to employ a matrix which has not been generated through Integration of RNA-ATAC, but following 10x RNA/ATAC pipeline. I therefore miss the integration matrix and so far I haven't got the chance to check for  eventual discrepancies in structure or naming with previous matrices (such as GeneExpression or GeneScore ones).


I am on this will be back to you soon!


Many thanks again for your help!


MZ


Matteo Zoia, Ph.D. Student
Gene Regulation & Cardiac Development
Osterwalder Lab
Department for BioMedical Research (DBMR)
University of Bern
Murtenstrasse 28, 369
CH - 3008 Bern
SWITZERLAND
________________________________
Da: NoemieL ***@***.***>
Inviato: martedì, 22. novembre 2022 18:09:53
A: GreenleafLab/ArchR
Cc: Zoia, Matteo (DBMR); Comment
Oggetto: Re: [GreenleafLab/ArchR] implementation of seurat DotPlot function (Discussion #882)


I looked in my data and your gene is not present in the GeneExpressionMatrix, I also tried the aliases BRGDA8, EIG18, and SSS2. I think this is the problem and the only way to get it is to find a way to not exclude its region on your scRNAseq data.

**Comment by matteozoia4** [comment]
Dear Noemie,


In the end the uppercase syntaxis was the issue, my bad I didn't looked at this in the immediate.

Many thanks for your time and additional explanations.


MZ


Matteo Zoia, Ph.D. Student
Gene Regulation & Cardiac Development
Osterwalder Lab
Department for BioMedical Research (DBMR)
University of Bern
Murtenstrasse 28, 369
CH - 3008 Bern
SWITZERLAND
________________________________
Da: Zoia, Matteo (DBMR)
Inviato: mercoledì, 23. novembre 2022 10:14:40
A: GreenleafLab/ArchR; GreenleafLab/ArchR
Cc: Comment
Oggetto: Re: [GreenleafLab/ArchR] implementation of seurat DotPlot function (Discussion #882)


Dear Noemie,


many thanks for your tempestive answer!

I am looking through this, I may indeed have set accidentally to uppercases the gene names, since working with mouse this could be definitely the issue for matching.

As well I need to employ a matrix which has not been generated through Integration of RNA-ATAC, but following 10x RNA/ATAC pipeline. I therefore miss the integration matrix and so far I haven't got the chance to check for  eventual discrepancies in structure or naming with previous matrices (such as GeneExpression or GeneScore ones).


I am on this will be back to you soon!


Many thanks again for your help!


MZ


Matteo Zoia, Ph.D. Student
Gene Regulation & Cardiac Development
Osterwalder Lab
Department for BioMedical Research (DBMR)
University of Bern
Murtenstrasse 28, 369
CH - 3008 Bern
SWITZERLAND
________________________________
Da: NoemieL ***@***.***>
Inviato: martedì, 22. novembre 2022 18:09:53
A: GreenleafLab/ArchR
Cc: Zoia, Matteo (DBMR); Comment
Oggetto: Re: [GreenleafLab/ArchR] implementation of seurat DotPlot function (Discussion #882)


I looked in my data and your gene is not present in the GeneExpressionMatrix, I also tried the aliases BRGDA8, EIG18, and SSS2. I think this is the problem and the only way to get it is to find a way to not exclude its region on your scRNAseq data.
